### ISSUE TITLE
GTK+-3.20: coco theme does not have round borders anymore

### ISSUE THREAD

**Issue by raveit65** [mate-desktop/mate-notification-daemon#98]
Coco theme does not have round borders anymore with gtk+-3.20

**Comment by monsta** [comment]
I see that in my VM with 3.18 as well.

**Comment by raveit65** [comment]
You're right, i see that with gtk+-3.16 too.
But it works well if marco use compositor with all gtk+ versions.

**Comment by monsta** [comment]
Yes, same here. Looks like it started in 3.16 as I don't see this issue with 3.14.

**Comment by raveit65** [comment]
Which m-n-d version are you using with gt+-3.14?
Maybe this is another result of infirits first re-work of m-n-d for gtk+-3 ?

**Comment by monsta** [comment]
Latest from git...

**Comment by monsta** [comment]
It seems that Coco actually has rounded borders, but w/o compositing, a black rectangle is painted around the notification window. Because Coco's background is black too, we see it as if the borders are not rounded. If you switch to Nodoka or Slider, you'll see the rounded corners and the black rectangle around the window.

**Comment by monsta** [comment]
So I thought it was happening only when the whole MATE is built with GTK+3... but it turned out it's not so. I've built the GTK+3 migration branch in GTK+2 environments with 3.16 and 3.18, and it still happens.

The issue looks similar to https://github.com/mate-desktop/marco/issues/163, maybe they're related.

@lukefromdc: can you please take a look at both issues? Currently I have no idea how to fix them...

**Comment by lukefromdc** [comment]
A transparent window is not possible without compositing with whatever
is underneath it. Round corners mean the drawn image is not using all
of the window, so without either a transparent window (composited) or
pseudotransparency(copying the background) you get black corners. 

I know pseudotransparency using cairo drawing works in GTK3 because it
works for custom panel backgrounds with or without compositing. I don't
know Cairo well enough to code that for the notification daemon themes
though.

I've never once seen transparency in notification themes work without 
composition in GTK3, so given that my GTK theme also requires it to work
with panel menus etc I had just considered this normal. This goes all the way
back to my first builds of 1.9, so my guess is that pseudotranparency code
used in GTK2 broke with the GTK3 transition.

Recall that in 1.9/1.10, most or all of the notification themes gave a totally
transparent background all the way across the notification window when 
compositing was used or black if it was not. The contents of these windows
have mostly been fixed but evidenly the windows themselves have not been,

**Comment by lukefromdc** [comment]
I have no idea how it is that with the compositor disabled you are ever getting rounded corners without an
underlying black background showing up as square "tips." Even less idea why this changes with where
the notification comes from, unless you have both mate-notification-daemon and another notification
daemon running, one but not both supporting pseudotransparency in which the background is read
and within the app composited into the final window.

**Comment by luigifab** [comment]
Another strange thing with GtkInspector:

![Image](https://github.com/user-attachments/assets/6923707f-d57b-4573-92cc-a080c7395f25)

When I use the "select an object" icon from titlebar... rounded corners are displayed!

![Image](https://github.com/user-attachments/assets/c3023218-4ff7-4cfe-81fd-5e8975189752)

Yes, compositor is disabled.
So I think there is only mate-notification-daemon.

**Comment by luigifab** [comment]
Found, this is `windata->composited`.

Anyway, there is a transparency without compositor, even if it's not the same when compositor is enabled.
There is a signal attached to the window: 

```c
g_signal_connect(win, "composited-changed", G_CALLBACK(on_composited_changed), windata);

static void on_composited_changed (GtkWidget* window, WindowData* windata)
{
	windata->composited = gdk_screen_is_composited (gtk_widget_get_screen (window));
	gtk_widget_queue_draw (window); // or windata->win
}
```

And when the notification window is created, there is some logic to set `windata->composited`:
```c
	windata->composited = FALSE;
	screen = gtk_window_get_screen(GTK_WINDOW(win));
	visual = gdk_screen_get_rgba_visual(screen);

	if (visual != NULL)
	{
		gtk_widget_set_visual(win, visual);
		if (gdk_screen_is_composited(screen))
			windata->composited = TRUE;
	}
```

But it's not always working.
When there is a black background, the variable is FALSE.

**Comment by lukefromdc** [comment]
What happens if you set that unconditionally to TRUE in the code? If pseudotransparency is not supported in the code (I have not looked), you should still get the black underlying background. If it is you should get the round corners every time but I do NOT know for sure what else may be involved.

**Comment by luigifab** [comment]
When I'm using `windata->composited = TRUE;` everywhere, there is no transparency without compositor (only black underlying background).

When I'm using `windata->composited = FALSE;` everywhere, there is transparency with and without compositor, but sometimes with black underlying background, and for example the hover with mouse on the close icon of nodoka theme "fix" the black underlying background!

I also tried to disable `composited-changed` event, but no...

**Comment by lukefromdc** [comment]
The mouse hover workaround triggering pseudotransparency when
the compositor is not used or the notification daemon is set to assume
it's not in use sounds like what's going on is a redraw issue, 
meaning the notification daemon
doesn't redraw the background until the mouse event is telling it to.
